lib/puppet/util/windows/registry.rb

```diff
@@ -207,21 +207,30 @@ def read(key, name_ptr, *rtype)
         # buffer is raw bytes, *not* chars - less a NULL terminator
         string_length = (byte_length / FFI.type_size(:wchar)) - 1 if byte_length > 0
 
-        case type
-        when Win32::Registry::REG_SZ, Win32::Registry::REG_EXPAND_SZ
-          result = [ type, data_ptr.read_wide_string(string_length) ]
-        when Win32::Registry::REG_MULTI_SZ
-          result = [ type, data_ptr.read_wide_string(string_length).split(/\0/) ]
-        when Win32::Registry::REG_BINARY
-          result = [ type, data.read_bytes(0, byte_length) ]
-        when Win32::Registry::REG_DWORD
-          result = [ type, data_ptr.read_dword ]
-        when Win32::Registry::REG_DWORD_BIG_ENDIAN
-          result = [ type, data_ptr.order(:big).read_dword ]
-        when Win32::Registry::REG_QWORD
-          result = [ type, data_ptr.read_qword ]
-        else
-          raise TypeError, "Type #{type} is not supported."
+        begin
+          case type
+            when Win32::Registry::REG_SZ, Win32::Registry::REG_EXPAND_SZ
+              result = [ type, data_ptr.read_wide_string(string_length) ]
+            when Win32::Registry::REG_MULTI_SZ
+              result = [ type, data_ptr.read_wide_string(string_length).split(/\0/) ]
+            when Win32::Registry::REG_BINARY
+              result = [ type, data_ptr.read_bytes(byte_length) ]
+            when Win32::Registry::REG_DWORD
+              result = [ type, data_ptr.read_dword ]
+            when Win32::Registry::REG_DWORD_BIG_ENDIAN
+              result = [ type, data_ptr.order(:big).read_dword ]
+            when Win32::Registry::REG_QWORD
+              result = [ type, data_ptr.read_qword ]
+            else
+              raise TypeError, "Type #{type} is not supported."
+          end
+        rescue IndexError => ex
+          if (ex.message =~ /^Memory access .* is out of bounds$/i)
+            parent_key_name = key.parent ? "#{key.parent.keyname}\\" : ""
+            Puppet.warning "A value in the registry key #{parent_key_name}#{key.keyname} is corrupt or invalid"
+          else
+            raise ex
```

Comment:
My thinking was to only handle specific error conditions in this low level function.  Errors should be handled by the publicly accessible methods like it is .open and .root

https://github.com/glennsarti/puppet/blob/ticket/stable/PUP-5591-5640-registry/lib/puppet/util/windows/registry.rb#L30-L32

However that is currently not the case.  Do you think I should be far more defensive about error handling in the other public methods?  Perhaps at least catch the errors and rethrow as a Puppet::Util::Windows::Error class?